crates/rpcsrv/src/session.rs

```diff
@@ -35,25 +36,49 @@ impl RemoteSession {
         session.get_session_catalog().get_state().as_ref().clone()
     }
 
-    pub async fn execute_serialized_plan(
+    pub async fn create_physical_plan(
         &self,
-        req: ExecuteRequest,
-    ) -> Result<SendableRecordBatchStream> {
-        match req.plan {
-            Some(Plan::LogicalPlan(buf)) => {
-                // TODO: Use a context that actually matters.
-                let fake_ctx = SessionContext::new();
-                let plan = LogicalPlanNode::try_decode(&buf)?
-                    .try_into_logical_plan(&fake_ctx, &GlareDBExtensionCodec)?;
-
-                let mut session = self.session.lock().await;
-                let physical = session.create_physical_plan(plan).await?;
-                let stream = session.execute_physical(physical)?;
-
-                Ok(stream)
-            }
-            Some(Plan::PhysicalPlan(_)) => Err(RpcsrvError::PhysicalPlansNotSupported),
-            None => Err(RpcsrvError::Internal("missing plan on request".to_string())),
-        }
+        logical_plan: impl AsRef<[u8]>,
+    ) -> Result<PhysicalPlanResponse> {
+        let mut session = self.session.lock().await;
+
+        // TODO: Use a context that actually matters.
+        let fake_ctx = SessionContext::new();
+        let plan = LogicalPlanNode::try_decode(logical_plan.as_ref())?
+            .try_into_logical_plan(&fake_ctx, &session.extension_codec())?;
+
+        let physical = session.create_physical_plan(plan).await?;
+        let schema = physical.schema();
+        let exec_id = session.add_physical_plan(physical);
+
+        Ok(PhysicalPlanResponse {
+            id: exec_id,
+            schema: Schema::clone(&schema),
+        })
+    }
+
+    pub async fn dispatch_access(
+        &self,
+        table_ref: OwnedTableReference,
+    ) -> Result<TableProviderResponse> {
+        let mut session = self.session.lock().await;
+        let provider = session.dispatch_access(table_ref).await?;
+        let schema = provider.schema();
+        let provider_id = session.add_table_provider(provider);
+
+        Ok(TableProviderResponse {
+            id: provider_id,
+            schema: Schema::clone(&schema),
+        })
+    }
+
+    pub async fn physical_plan_execute(&self, exec_id: Uuid) -> Result<SendableRecordBatchStream> {
```

Comment:
I'm worried that this stateful pattern of `add/get` physical plan may cause some contention as we scale up, or run multiple clients. 

What issues were there with the remoteExec node? It seems to fit better with datafusion. couldn't we just setup the extensionCodec with a registry of table providers?

Reply:
The issue is serializing table providers and physical plans. It's not as simple. We'll have to change a lot to make them serializable. For example, how do you serialize a provider with postgres connection? We either need the connection on the server (stateful again), or we create a new connection everytime.

Reply:
Yeah, we do have physical plan extension similar to logical plans. I did think about this but at this moment, getting everything there requires a lot of effort. But we should definitely go there in the future.